<p style="text-align:center">
    <a href="https://skills.network" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **Exception Handling**

Estimated time needed: **15** minutes
    

## Objectives

After completing this lab you will be able to:

* Understand exceptions    
* Handle the exceptions


## Table of Contents


* What is an Exception?
* Exception Handling


* Qu’est-ce qu’une exception ?
* Gestion des exceptions


----


## What is an Exception?


In this section you will learn about what an exception is and see examples of them.

Dans cette section, vous apprendrez ce qu’est une exception et en verrez des exemples.

### Definition


An exception is an error that occurs during the execution of code. This error causes the code to raise an exception and if not prepared to handle it will halt the execution of the code.




Une exception est une erreur qui se produit lors de l’exécution du code. Cette erreur provoque le déclenchement d’une exception par le code et, s’il n’est pas prêt à la gérer, l’exécution du code est interrompue.

### Examples


Run each piece of code and observe the exception raised


Exécutez chaque morceau de code et respectez l’exception levée

In [1]:
1/0

ZeroDivisionError: division by zero

<code>ZeroDivisionError</code> occurs when you try to divide by zero.

<code>ZeroDivisionError</code> se produit lorsque vous essayez de diviser par zéro.

In [2]:
y = a + 5

NameError: name 'a' is not defined

<code>NameError</code> -- in this case, it means that you tried to use the variable a when it was not defined.

<code>NameError</code> -- dans ce cas, cela signifie que vous avez essayé d’utiliser la variable a alors qu’elle n’était pas définie.

In [3]:
a = [1, 2, 3]
a[10]

IndexError: list index out of range

<code>IndexError</code> -- in this case, it occured because you tried to access data from a list using an index that does not exist for this list.


<code>IndexError</code> : dans ce cas, cela s’est produit parce que vous avez essayé d’accéder aux données d’une liste à l’aide d’un index qui n’existe pas pour cette liste.

In [4]:
a[1]

2

In [5]:
a[2]

3

There are many more exceptions that are built into Python, here is a list of them https://docs.python.org/3/library/exceptions.html


Il existe de nombreuses autres exceptions intégrées à Python, en voici une liste https://docs.python.org/3/library/exceptions.html

## Exception Handling

## Gestion des exceptions

In this section you will learn how to handle exceptions. You will understand how to make your program perform specified tasks instead of halting code execution when an exception is encountered.



Dans cette section, vous apprendrez à gérer les exceptions. Vous comprendrez comment faire en sorte que votre programme exécute des tâches spécifiées au lieu d’arrêter l’exécution du code lorsqu’une exception est rencontrée.


### Try Except

### Essayez sauf

A <code>try except</code> will allow you to execute code that might raise an exception and in the case of any exception or a specific one we can handle or catch the exception and execute specific code. This will allow us to continue the execution of our program even if there is an exception.

Python tries to execute the code in the <code>try</code> block. In this case if there is any exception raised by the code in the <code>try</code> block, it will be caught and the code block in the <code>except</code> block will be executed. After that, the code that comes <em>after</em> the try except will be executed.


Un <code>essai sauf</code> vous permettra d’exécuter du code qui pourrait lever une exception et dans le cas d’une exception ou d’une exception spécifique, nous pouvons gérer ou attraper l’exception et exécuter du code spécifique. Cela nous permettra de poursuivre l’exécution de notre programme même s’il y a une exception.

Python essaie d’exécuter le code dans le bloc <code>try</code>. Dans ce cas, s’il y a une exception levée par le code dans le bloc <code>try</code>, elle sera interceptée et le bloc de code dans le bloc <code>except</code> sera exécuté. Après cela, le code qui vient <em>après</em> l’essai except sera exécuté.


In [6]:
# potential code before try catch
# code potentiel avant try catch

try:
    # code to try to execute
    # code à essayer d’exécuter
except:
    # code to execute if there is an exception
    # code à exécuter s’il y a une exception
    
# code that will still execute if there is an exception
# code qui s’exécutera toujours s’il y a une exception


IndentationError: expected an indented block (4107979928.py, line 5)

### Try Except Example


In this example we are trying to divide a number given by the user, save the outcome in the variable <code>a</code>, and then we would like to print the result of the operation. When taking user input and dividing a number by it there are a couple of exceptions that can be raised. For example if we divide by zero. Try running the following block of code with <code>b</code> as a number. An exception will only be raised if <code>b</code> is zero.


Dans cet exemple, nous essayons de diviser un nombre donné par l’utilisateur, d’enregistrer le résultat dans la variable <code>a</code>, puis d’imprimer le résultat de l’opération. Lorsque vous prenez une entrée utilisateur et divisez un nombre par celle-ci, il y a quelques exceptions qui peuvent être soulevées. Par exemple, si nous divisons par zéro.Essayez d’exécuter le bloc de code suivant avec <code>b</code> comme nombre. Une exception ne sera levée que si <code>b</code> est égal à zéro.


In [7]:
a = 1

try:
    b = int(input("Please enter a number to divide a"))
    a = a/b
    print("Success a=",a)
except:
    print("There was an error")
        


Please enter a number to divide a 0


There was an error


### Try Except Specific

### Essayez sauf spécifique

A specific <code>try except</code> allows you to catch certain exceptions and also execute certain code depending on the exception. This is useful if you do not want to deal with some exceptions and the execution should halt. It can also help you find errors in your code that you might not be aware of. Furthermore, it can help you differentiate responses to different exceptions. In this case, the code after the try except might not run depending on the error.


Un <code>try except</code> spécifique vous permet d’intercepter certaines exceptions et d’exécuter du code en fonction de l’exception. Ceci est utile si vous ne souhaitez pas traiter certaines exceptions et que l’exécution doit s’arrêter. Cela peut également vous aider à trouver des erreurs dans votre code dont vous n’êtes peut-être pas conscient. En outre, il peut vous aider à différencier les réponses aux différentes exceptions. Dans ce cas, le code après try except peut ne pas s’exécuter en fonction de l’erreur.


<b>Do not run, just to illustrate:</b>

<b>Ne courez pas, juste pour illustrer :</b>

In [ ]:
# potential code before try catch

try:
    # code to try to execute
except (ZeroDivisionError, NameError):
    # code to execute if there is an exception of the given types
    # code à exécuter s’il y a une exception des types donnés
    
# code that will execute if there is no exception or a one that we are handling
# code qui s’exécutera s’il n’y a pas d’exception ou une exception que nous gérons

In [ ]:
# potential code before try catch

try:
    # code to try to execute
except ZeroDivisionError:
    
    # code to execute if there is a ZeroDivisionError
    # code à exécuter s’il y a une ZeroDivisionError
    
except NameError:
    # code to execute if there is a NameError
    # code à exécuter s’il y a une NameError
    
# code that will execute if there is no exception or a one that we are handling

You can also have an empty <code>except</code> at the end to catch an unexpected exception:

Vous pouvez également avoir un <code>vide sauf</code> à la fine pour attraper une exception inattendue :

<b>Do not run, just to illustrate:</b>


In [ ]:
# potential code before try catch

try:
    # code to try to execute
except ZeroDivisionError:
    # code to execute if there is a ZeroDivisionError
except NameError:
    # code to execute if there is a NameError
except:
    # code to execute if ther is any exception
    # code à exécuter s’il y a une exception
    
# code that will execute if there is no exception or a one that we are handling

### Try Except Specific Example

### Essayez sauf exemple spécifique

This is the same example as above, but now we will add differentiated messages depending on the exception, letting the user know what is wrong with the input.



C’est le même exemple que ci-dessus, mais nous allons maintenant ajouter des messages différenciés en fonction de l’exception, permettant à l’utilisateur de savoir ce qui ne va pas avec l’entrée.

In [ ]:
a = 1

try:
    b = int(input("Please enter a number to divide a"))
    a = a/b
    print("Success a=",a)
except ZeroDivisionError:
    print("The number you provided cant divide 1 because it is 0, Le nombre que vous avez fourni ne peut pas diviser 1 car il est égal à 0")
except ValueError:
    print("You did not provide a number, Vous n’avez pas fourni de numéro")
except:
    print("Something went wrong, Quelque chose s’est mal passé ")
        


### Try Except Else and Finally


<code>else</code> allows one to check if there was no exception when executing the try block. This is useful when we want to execute something only if there were no errors.

<code>else</code> permet de vérifier s’il n’y a pas eu d’exception lors de l’exécution du bloc try. Ceci est utile lorsque nous voulons exécuter quelque chose uniquement s’il n’y a pas eu d’erreurs.


<b>do not run, just to illustrate</b>


In [ ]:
# potential code before try catch

try:
    # code to try to execute
except ZeroDivisionError:
    # code to execute if there is a ZeroDivisionError
except NameError:
    # code to execute if there is a NameError
except:
    # code to execute if ther is any exception
else:
    # code to execute if there is no exception
    # code à exécuter s’il n’y a pas d’exception
    
# code that will execute if there is no exception or a one that we are handling
# code qui s’exécutera s’il n’y a pas d’exception ou une exception que nous gérons

<code>finally</code> allows us to always execute something even if there is an exception or not. This is usually used to signify the end of the try except.



<code>nous permet enfin</code> de toujours exécuter quelque chose même s’il y a une exception ou non. Ceci est généralement utilisé pour signifier la fin de l’essai, sauf.

In [ ]:
# potential code before try catch

try:
    # code to try to execute
except ZeroDivisionError:
    # code to execute if there is a ZeroDivisionError
except NameError:
    # code to execute if there is a NameError
except:
    # code to execute if ther is any exception
else:
    # code to execute if there is no exception
finally:
    # code to execute at the end of the try except no matter what
     # code à exécuter à la fin de l’essai, mais quoi qu’il arrive
    
# code that will execute if there is no exception or a one that we are handling

### Try Except Else and Finally Example


You might have noticed that even if there is an error the value of <code>a</code> is always printed. Let's use the <code>else</code> and print the value of <code>a</code> only if there is no error.



Vous avez peut-être remarqué que même en cas d’erreur, la valeur de <code>a</code> est toujours affichée. Utilisons <code>else et</code> affichons la valeur de <code>a</code> uniquement s’il n’y a pas d’erreur.

In [ ]:
a = 1

try:
    b = int(input("Please enter a number to divide a"))
    a = a/b
except ZeroDivisionError:
    print("The number you provided cant divide 1 because it is 0")
except ValueError:
    print("You did not provide a number")
except:
    print("Something went wrong")
else:
    print("success a=",a)

Now lets let the user know that we are done processing their answer. Using the <code>finally</code>, let's add a print.


Faisons maintenant savoir à l’utilisateur que nous avons fini de traiter sa réponse. En utilisant le <code>enfin</code>, ajoutons une impression.

In [ ]:
a = 1

try:
    b = int(input("Please enter a number to divide a"))
    a = a/b
except ZeroDivisionError:
    print("The number you provided cant divide 1 because it is 0")
except ValueError:
    print("You did not provide a number")
except:
    print("Something went wrong")
else:
    print("success a=",a)
finally:
    print("Processing Complete Traitement terminé")

<center>
    
    
# Practice Exercises 

</center>


## Exercise 1: Handling ZeroDivisionError


Imagine you have two numbers and want to determine what happens when you divide one number by the other. To do this, you need to create a Python function called `safe_divide.` You give this function two numbers, a `'numerator'` and a `'denominator'`. The 'numerator' is the number you want to divide, and the `'denominator'` is the number you want to divide by. Use the user input method of Python to take the values.

The function should be able to do the division for you and give you the result. But here's the catch: if you try to divide by zero (which is not allowed in math), the function should be smart enough to catch that and tell you that it's not possible to divide by zero. Instead of showing an error, it should return None, which means 'nothing' or 'no value', and print `"Error: Cannot divide by Zero.`


Imaginez que vous avez deux nombres et que vous voulez déterminer ce qui se passe lorsque vous divisez un nombre par l’autre. Pour ce faire, vous devez créer une fonction Python appelée 'safe_divide'', à laquelle vous donnez deux nombres, un ''numérateur'' et un ''dénominateur''. Le « numérateur » est le nombre que vous voulez diviser, et le « dénominateur » est le nombre par lequel vous voulez diviser. Utilisez la méthode d’entrée utilisateur de Python pour prendre les valeurs.

La fonction doit être capable de faire la division pour vous et de vous donner le résultat. Mais voici le hic : si vous essayez de diviser par zéro (ce qui n’est pas autorisé en mathématiques), la fonction devrait être assez intelligente pour l’attraper et vous dire qu’il n’est pas possible de diviser par zéro. Au lieu d’afficher une erreur, il devrait renvoyer None, ce qui signifie « rien » ou « aucune valeur », et afficher « Erreur : Impossible de diviser par zéro ».

Cellule non exécutée en raison d’une entrée en attente
La cellule n’a pas été exécutée pour éviter l’interblocage du noyau car il y a une autre entrée en attente ! Soumettez votre contribution en attente et réessayez.



In [14]:
#Type your code here

def safe_divide(numerator, denominator):
    try:
        result = numerator/denominator
        return result
        
    except ZeroDivisionError:
        print("Error: cannot divide by zero")
        return None
        
# Test case
numerator= int(input("Enter the numerator value:-"))

denominator= int(input("Enter the denominator value:-"))
print(safe_divide(numerator,denominator))

Enter the numerator value:- 15
Enter the denominator value:- 0


Error: cannot divide by zero
None


In [ ]:
def safe_divide(numerator, denominator):
    try:
        result = 15/0
        return result
        
    except ZeroDivisionError:
        print("Error: cannot divide by zero")
        return None
        
# Test case
numerator= int(input("Enter the numerator value:-"))

denominator= int(input("Enter the denominator value:-"))
print(safe_divide(numerator,denominator))

**Note:- Practice handling exceptions by trying different input types like using integers, strings, zero, negative values, or other data types.**

**Remarque :- Entraînez-vous à gérer les exceptions en essayant différents types d’entrée tels que l’utilisation d’entiers, de chaînes, de zéro, de valeurs négatives ou d’autres types de données.**

<details><summary>Click here for the hints</summary>

```python
Follow these: -
* Define a function to perform the division and use two arguments.
* Use the try-except block to handle ZeroDivisionError.
* Return the result of the division if no error occurs.
* Return None if division by zero occurs.
* take user input for numerator and denominator values.
* call and print your function with the user inputs.
**Note:- Test with different inputs to validate error handling.**
```

</details>


<details><summary>Cliquez ici pour les conseils</summary>

'''python
Suivez ceux-ci : -
* Définissez une fonction pour effectuer la division et utilisez deux arguments.
* Utilisez le bloc try-except pour gérer ZeroDivisionError.
* Renvoyer le résultat de la division si aucune erreur ne se produit.
* Renvoie Aucun si la division par zéro se produit.
* Prendre les entrées de l’utilisateur pour les valeurs du numérateur et du dénominateur.
* Appelez et imprimez votre fonction avec les entrées de l’utilisateur.
**Remarque :- Test avec différentes entrées pour valider la gestion des erreurs.**
```

</details>


<details><summary>Click here for the solution</summary>

```python

def safe_divide(numerator,denominator):
    try:
        result = numerator / denominator
        return result
    except ZeroDivisionError:
        print("Error: Cannot divide by zero.")
        return None
# Test case
numerator=int(input("Enter the numerator value:-"))
denominator=int(input("Enter the denominator value:-"))
print(safe_divide(numerator,denominator)) 
```

</details>


## Exercise 2: Handling ValueError


Imagine you have a number and want to calculate its square root. To do this, you need to create a Python function. You give this function one number, `'number1'`.

The function should generate the square root value if you provide a positive integer or float value as input. However, the function should be clever enough to detect the mistake if you enter a negative value. It should kindly inform you with a message saying, `'Invalid input! Please enter a positive integer or a float value.`


Imaginez que vous avez un nombre et que vous voulez calculer sa racine carrée. Pour ce faire, vous devez créer une fonction Python. Vous donnez à cette fonction un chiffre, ''number1''.

La fonction doit générer la valeur de la racine carrée si vous fournissez un entier positif ou une valeur flottante en entrée. Cependant, la fonction doit être suffisamment intelligente pour détecter l’erreur si vous entrez une valeur négative. Il devrait gentiment vous informer avec un message disant : ''Entrée invalide ! Veuillez saisir un entier positif ou une valeur flottante.

In [21]:
#Type your code here

import math


def perform_calculation(number1):
    try:
        result= math.sqrt(number1)
        print(f"RESULT: {result}")   
    except ValueError:
        print("Invalid input! Please enter a positive integer or a float value.")
        
# Test case
number1=int(input("Enter the number:-"))
perform_calculation(number1)

Enter the number:- 5


RESULT: 2.23606797749979


**Note:- Practice handling exceptions by trying different input types like using integers, strings, zero, negative values, or other data types.**


**Remarque :- Entraînez-vous à gérer les exceptions en essayant différents types d’entrée tels que l’utilisation d’entiers, de chaînes, de zéro, de valeurs négatives ou d’autres types de données.**

<details><summary>Cliquez ici pour les conseils</summary>

'''python
Suivez ces éléments :
* Définissez une fonction pour exécuter la racine carrée de l’argument. 
* Utilisez le bloc try-except pour la gestion des erreurs.
* Utilisez la fonction 'sqrt()' du package 'math' pour calculer la racine carrée. Attrapez ValueError et affichez le message d’erreur."
* Prenez l’entrée utilisateur de la valeur, number1. 
* Test avec des nombres négatifs pour valider la gestion des erreurs.
```

</details>

<details><summary>Click here for the hints</summary>

```python
Follow these:-
* Define a function to perform square root of the argument. 
* Use try-except block for error handling.
* Use `sqrt()` function from the `math` package to calculate the square root. Catch ValueError and display the error message."
* Take user input of the value, number1. 
* Test with negative numbers to validate error handling.
```

</details>


<details><summary>Click here for the solution</summary>

```python
import math

def perform_calculation(number1):
    try:
        result = math.sqrt(number1)
        print(f"Result: {result}")
    except ValueError:
        print("Error: Invalid input! Please enter a positive integer or a float value.")
# Test case
number1=float(input("Enter the number:-"))
perform_calculation(number1)
```

**Note:- Test with different inputs to validate error handling.**

</details>


## Exercise 3: Handling Generic Exceptions


## Exercice 3 : Gestion des exceptions génériques

Imagine you have a number and want to perform a complex mathematical task. The calculation requires dividing the value of the input argument `"num"` by the difference between `"num"` and 5, and the result has to be stored in a variable called `"result"`.

You have to define a function so that it can perform that complex mathematical task. The function should handle any potential errors that occur during the calculation. To do this, you can use a try-except block. If any exception arises during the calculation, it should catch the error using the generic exception class `"Exception" as "e"`. When an exception occurs, the function should display `"An error occurred during calculation.`


Imaginez que vous avez un nombre et que vous voulez effectuer une tâche mathématique complexe. Le calcul nécessite de diviser la valeur de l’argument d’entrée ''num"' par la différence entre ''num"' et 5, et le résultat doit être stocké dans une variable appelée ''result"'.


Vous devez définir une fonction pour qu’elle puisse effectuer cette tâche mathématique complexe. La fonction doit gérer toutes les erreurs potentielles qui se produisent pendant le calcul. Pour ce faire, vous pouvez utiliser un bloc try-except. Si une exception survient pendant le calcul, il doit intercepter l’erreur en utilisant la classe d’exception générique '"Exception » comme « e"'. Lorsqu’une exception se produit, la fonction doit afficher « Une erreur s’est produite pendant le calcul ».

In [22]:
def complex_calculation(num):
    try:
        result= num/(num-5)
        
        print(f"Result: {result}")
        
    except Exception as e:
        print("An error occurred during calculation.")
#Test case 
user_input=float(input("Enter a number:"))
complex_calculation(user_input)

Enter a number: 1


Result: -0.25


In [ ]:
#Type your code here



**Note:- Practice handling exceptions by trying different input types like using integers, strings, zero, negative values, or other data types.**


détails><summary>Cliquez ici pour les conseils</summary>

'''python
Suivez ces éléments :
* Définissez une fonction pour le calcul complexe et passez n’importe quel argument.
* Utilisez un bloc try-except pour la gestion des erreurs.
* Effectuez le calcul et stockez le résultat dans « résultat ».
* Interceptez toutes les exceptions en utilisant Exception comme e.
* Affiche « Une erreur s’est produite lors du calcul. » lorsqu’une exception est interceptée.
* prendre en compte les entrées de l’utilisateur
* Appeler la fonction définie avec l’entrée utilisateur
**Remarque :- Test avec différentes entrées pour valider la gestion des erreurs.**
```

</details>


<details><summary>Click here for the hints</summary>

```python
Follow these:-
* Define a function for the complex calculation and pass any argument.
* Use a try-except block for error handling.
* Perform the calculation and store the result in "result."
* Catch any exceptions using Exception as e.
* Display "An error occurred during calculation." when an exception is caught.
* take user input
* Call the defined function with the user input
**Note:- Test with different inputs to validate error handling.**
```

</details>


<details><summary>Click here for the solution</summary>

```python
def complex_calculation(num):
    try:
        result = num / (num - 5)
        print (f"Result: {result}")
    except Exception as e:
        print("An error occurred during calculation.")
# Test case
user_input = float(input("Enter a number: "))
complex_calculation(user_input)
```

</details>


## Authors


<a href="https://www.linkedin.com/in/joseph-s-50398b136/" target="_blank">Joseph Santarcangelo</a>


## Change Log

|  Date (YYYY-MM-DD) |  Version | Changed By  |  Change Description |
|---|---|---|---|
| 2023-11-02 | 2.2 | Abhishek Gagneja | Updated instructions |
| 2023-08-01 | 2.1   | Akansha Yadav | Added optional practice section |
| 2020-09-02  | 2.0  | Simran | Template updates to the file|
|   |   |   |   |




## <h3 align="center"> © IBM Corporation 2023. All rights reserved. <h3/>
